### PR TITLE
enhance CheckWhenObserve AnsibleRunPolicy

### PR DESCRIPTION
**Pull request by fahedouch**
Signed-off-by: Fahed DORGAA <fahed.dorgaa@gmail.com>

### Description of your changes

Fixes # Enhancing of CheckWhenObserve Policy

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: internal/controller/ansibleRun/ansibleRun.go [hunk 1]
```diff
@@ -310,7 +314,24 @@ func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
 
 		return c.handleLastApplied(last, cr)
 	case "CheckWhenObserve":
-		// TODO
+		dc, err := c.runner.Run(true)
```
Comment (morningspace) [inline internal/controller/ansibleRun/ansibleRun.go:317]: It's a bit strange to have a `true|false` value passed in every time when calling the Ansible Runner encapsulated go method. How about using default parameter or the pattern such as `WithCheckMode()`?

**Review by morningspace — COMMENTED** [review]
@fahedouch I finished the review and added a few comments. Besides, can we create an example under examples folder just to demonstrate the use of this policy?

File: internal/ansible/ansible.go [hunk 2]
```diff
@@ -138,17 +139,20 @@ func withAnsibleRunPolicy(p *RunPolicy) runnerOption {
...
 func (p Parameters) playbookCmdFunc(ctx context.Context, path string) cmdFuncType {
-	return func(_ string, hosts string, verbosity int) *exec.Cmd {
+	return func(_ string, hosts string, verbosity int, checkMode bool) *exec.Cmd {
 		cmdArgs := []string{"run", p.WorkingDirPath}
 		cmdOptions := []string{
 			"-p", path,
 			"--hosts", hosts,
 		}
 
+		if checkMode {
```
[the remark this answers is not in the document]
Comment (fahedouch) [reply internal/ansible/ansible.go:153]: fixed

File: internal/controller/ansibleRun/ansibleRun.go [hunk 3]
```diff
@@ -319,31 +340,35 @@ func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.Ex
...
+	if err = dc.Wait(); err != nil {
+		return managed.ExternalCreation{}, err
+	}
 
-	return managed.ExternalCreation{
-		// Optionally return any details that may be required to connect to the
-		// external resource. These will be stored as the connection secret.
-		ConnectionDetails: managed.ConnectionDetails{},
-	}, nil
+	return managed.ExternalCreation{}, nil
```
[the remark this answers is not in the document]
Comment (fahedouch) [reply internal/controller/ansibleRun/ansibleRun.go]: fixed

File: internal/controller/ansibleRun/ansibleRun.go (hunk 1 above)
Comment (fahedouch) [reply internal/controller/ansibleRun/ansibleRun.go]: fixed

**Comment by fahedouch** [comment]
@morningspace  This PR leverage multiple fixes:

- fix showing inline ansible content output in logs
- fixing API
- controller refacto
- refacto test examples

File: apis/v1alpha1/ansibleRun_types.go [hunk 4]
```diff
@@ -47,15 +45,16 @@ type AnsibleRunParameters struct {
...
 
 	// The remote configuration of this AnsibleRun; the content can be retrieved from Ansible Galaxy as community contents
 	// +optional
 	Playbooks []string `json:"playbooks"`
 
 	// Configuration variables.
+	// +kubebuilder:pruning:PreserveUnknownFields
 	// +optional
-	Vars []Var `json:"vars,omitempty"`
+	Vars runtime.RawExtension `json:"vars,omitempty"`
```
[the remark this answers is not in the document]
Comment (fahedouch) [reply apis/v1alpha1/ansibleRun_types.go:51]: fixed

**Review by morningspace — APPROVED** [review]

File: internal/ansible/ansible.go [hunk 5]
```diff
@@ -191,20 +203,21 @@ func (p Parameters) roleCmdFunc(ctx context.Context, path string) (cmdFuncType,
...
 }
 
 // GalaxyInstall Install non-exists collections/roles with ansible-galaxy cli
-func (p Parameters) GalaxyInstall(ctx context.Context, behaviorVars map[string]string, isCollectionRequirements, isRoleRequirements bool) error {
+func (p Parameters) GalaxyInstall(ctx context.Context, behaviorVars map[string]string, requirementsType string) error {
 	requirementsFilePath := runnerutil.GetFullPath(p.WorkingDirPath, galaxyutil.RequirementsFile)
 	var cmdArgs []string
 	var cmdOptions []string
-	if isCollectionRequirements {
+	switch requirementsType {
```
Comment (morningspace) [inline internal/ansible/ansible.go:214]: I just created a new issue at https://github.com/multicloudlab/crossplane-provider-ansible/issues/81 as an enhancement in future release, as it looks here the requirements type are mutually exclusive for `roles` and `collections`. It seems Ansible requirements.yml natively support mixed references to both. Let me know if that makes sense or not?

Comment (fahedouch) [reply internal/ansible/ansible.go:214]: thanks

File: examples/ansible/ansibleRun-checkWhenObserve-policy.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    name: gcpconfig
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  namespace: crossplane-system
+  name: gcp-credentials
+type: Opaque
+data:
+  credentials: QkFTRTY0RU5DT0RFRF9QUk9WSURFUl9DUkVEUw==
```
[the remark this answers is not in the document]
Comment (fahedouch) [reply examples/ansible/ansibleRun-checkWhenObserve-policy.yml]: fixed

**Pull request merged**